apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/test/java/gov/cms/bfd/pipeline/rda/grpc/sink/FissClaimRdaSinkTest.java

```diff
@@ -0,0 +1,175 @@
+    doThrow(new EntityExistsException()).when(entityManager).persist(batch.get(0));
+    doThrow(new EntityNotFoundException()).when(entityManager).merge(batch.get(1));
+
+    try {
+      sink.writeBatch(batch);
+      fail("should have thrown");
+    } catch (ProcessingException error) {
+      assertEquals(0, error.getProcessedCount());
+      assertThat(error.getCause(), CoreMatchers.instanceOf(EntityNotFoundException.class));
+    }
+
+    verify(entityManager).persist(batch.get(0));
+    verify(entityManager, times(0)).persist(batch.get(1)); // not called once a persist fails
+    verify(entityManager, times(0)).persist(batch.get(2)); // not called once a persist fails
+    verify(entityManager).merge(batch.get(0));
+    verify(entityManager).merge(batch.get(1));
+    verify(entityManager, times(0)).persist(batch.get(2)); // not called once a merge fails
+    verify(transaction, times(2)).rollback(); // both persist and merge transactions are rolled back
+
+    assertMeterReading(1, FissClaimRdaSink.CALLS_METER_NAME);
+    assertMeterReading(0, FissClaimRdaSink.OBJECTS_PERSISTED_METER_NAME);
+    assertMeterReading(0, FissClaimRdaSink.OBJECTS_MERGED_METER_NAME);
+    assertMeterReading(0, FissClaimRdaSink.OBJECTS_WRITTEN_METER_NAME);
+    assertMeterReading(1, FissClaimRdaSink.FAILURES_METER_NAME);
+  }
+
+  @Test
+  public void persistFatalError() {
+    final List<PreAdjFissClaim> batch =
+        ImmutableList.of(createClaim("1"), createClaim("2"), createClaim("3"));
+    doThrow(new RuntimeException("oops")).when(entityManager).persist(batch.get(1));
+
+    try {
+      sink.writeBatch(batch);
+      fail("should have thrown");
+    } catch (ProcessingException error) {
+      assertEquals(0, error.getProcessedCount());
+      assertThat(error.getCause(), CoreMatchers.instanceOf(RuntimeException.class));
+    }
+
+    verify(entityManager).persist(batch.get(0));
+    verify(entityManager).persist(batch.get(1));
+    verify(entityManager, times(0)).persist(batch.get(2)); // not called once a persist fails
+    verify(entityManager, times(0))
+        .merge(any()); // non-duplicate key error prevents any calls to merge
+    verify(transaction).rollback();
+
+    assertMeterReading(1, FissClaimRdaSink.CALLS_METER_NAME);
+    assertMeterReading(0, FissClaimRdaSink.OBJECTS_PERSISTED_METER_NAME);
+    assertMeterReading(0, FissClaimRdaSink.OBJECTS_MERGED_METER_NAME);
+    assertMeterReading(0, FissClaimRdaSink.OBJECTS_WRITTEN_METER_NAME);
+    assertMeterReading(1, FissClaimRdaSink.FAILURES_METER_NAME);
+  }
+
+  @Test
+  public void closeMethodsAreCalled() throws Exception {
+    sink.close();
+    verify(dataSource).close();
+    verify(entityManager).close();
+    verify(entityManagerFactory).close();
+  }
+
+  @Test
+  public void exceptionMessageDuplicateKeyDetection() {
+    assertEquals(
+        true, isDuplicateKeyException(new RuntimeException(new IOException("key already existS"))));
```

Comment:
```suggestion
        true, isDuplicateKeyException(new RuntimeException(new IOException("key already exists"))));
```

Reply:
Thanks, it wasn't obvious but actually that was deliberate to confirm the string comparison was case insensitive.